File: base/src/main/java/com/xjek/base/base/BaseFragment.java
package com.xjek.base.base;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.xjek.base.views.CustomDialog;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

public abstract class BaseFragment<T extends ViewDataBinding> extends Fragment {

    private FragmentActivity mActivity;
    private T mViewDataBinding;
    private MutableLiveData loadingLiveData = new MutableLiveData<Boolean>();
    private CustomDialog customDialog;


    @LayoutRes
    public abstract int getLayoutId();

    protected abstract void initView(View mRootView, ViewDataBinding mViewDataBinding);

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mActivity = getActivity();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mViewDataBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);
        return mViewDataBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView(mViewDataBinding.getRoot(), mViewDataBinding);
        customDialog = new CustomDialog(mActivity);
        loadingLiveData.observe(getActivity(), new Observer() {
            @Override
            public void onChanged(Object o) {
                /*if(loadingLiveData.getValue()==true){
                    showLoading();
                }else{
                    hideLoading();
                }*/
            }
        });

    }

    protected void setBindingVariable(int variableId, @Nullable Object object) {
        mViewDataBinding.setVariable(variableId, object);
        mViewDataBinding.executePendingBindings();
    }

    protected void launchNewActivity(Class<?> cls, boolean shouldCloseActivity) {
        startActivity(new Intent(mActivity.getApplicationContext(), cls));
        if (shouldCloseActivity)
            mActivity.finish();
    }

    @Override
    public void onDetach() {
        mActivity = null;
        super.onDetach();
    }

    protected void showLoading() {
        if (customDialog.getWindow() != null) {
            customDialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
            customDialog.show();
        }
    }

    protected void hideLoading() {
        customDialog.cancel();
    }


}

File: base/src/main/java/com/gox/base/base/BaseFragment.kt
package com.gox.base.base

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.MutableLiveData
import com.gox.base.extensions.observeLiveData
import com.gox.base.utils.LocaleUtils
import com.gox.base.views.CustomDialog

abstract class BaseFragment<T : ViewDataBinding> : Fragment() {

    private var mActivity: FragmentActivity? = null
    private lateinit var mViewDataBinding: T
    private val loadingLiveData = MutableLiveData<Boolean>()
    private var customDialog: CustomDialog? = null

    val loadingObservable: MutableLiveData<*>
        get() = loadingLiveData

    @LayoutRes
    protected abstract fun getLayoutId(): Int

    protected abstract fun initView(mRootView: View?, mViewDataBinding: ViewDataBinding?)

    override fun onAttach(context: Context) {
        super.onAttach(context)
        LocaleUtils.setLocale(context)
        mActivity = activity
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        mViewDataBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false)
        return mViewDataBinding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView(mViewDataBinding.root, mViewDataBinding)
        customDialog = CustomDialog(mActivity!!,true)

        observeLiveData(loadingLiveData) { isShowLoading ->
            if (isShowLoading) showLoading() else hideLoading()
        }
    }

    protected fun setBindingVariable(variableId: Int, value: Any) {
        mViewDataBinding.setVariable(variableId, value)
        mViewDataBinding.executePendingBindings()
    }

    protected fun launchNewActivity(cls: Class<*>, shouldCloseActivity: Boolean) {
        startActivity(Intent(mActivity!!.applicationContext, cls))
        if (shouldCloseActivity)
            mActivity!!.finish()
    }

    override fun onDetach() {
        mActivity = null
        super.onDetach()
    }

    protected fun showLoading() {
        if (customDialog!!.window != null) {
            customDialog!!.window!!.setBackgroundDrawableResource(android.R.color.transparent)
            customDialog!!.show()
        }
    }

    protected fun hideLoading() {
        customDialog!!.cancel()
    }

}